File: src/main/java/net/gtaun/shoebill/common/command/CommandEntryInternal.java
package net.gtaun.shoebill.common.command;

import net.gtaun.shoebill.object.Player;
import org.apache.commons.lang3.StringUtils;

import java.util.List;

class CommandEntryInternal {
    public static String completePath(String path, String child) {
        return StringUtils.isBlank(path) ? child : path + " " + child;
    }

    @FunctionalInterface
    public interface CommandHandlerInternal {
        boolean handle(Player player, Object[] params);
    }


    private String command;
    private Class<?>[] paramTypes;
    private short priority;
    private String helpMessage;
    private CommandHandlerInternal handler;
    private boolean caseSensitive;
    private String categorie;
    private List<CustomCommandHandler> beforeCheckers;
    private List<CustomCommandHandler> customHandlers;
    private CommandParameter[] parameterAnnotations;

    public CommandEntryInternal(String command, Class<?>[] paramTypes, short priority, String helpMessage, boolean caseSensitive,
                                CommandHandlerInternal handler, String categorie, List<CustomCommandHandler> beforeCheckers, List<CustomCommandHandler> customHandlers,
                                CommandParameter[] parameterAnnotations) {
        if(parameterAnnotations == null) parameterAnnotations = new CommandParameter[0];
        this.command = command;
        this.paramTypes = paramTypes;
        this.priority = priority;
        this.helpMessage = helpMessage;
        this.caseSensitive = caseSensitive;
        this.handler = handler;
        this.categorie = categorie;
        this.beforeCheckers = beforeCheckers;
        this.customHandlers = customHandlers;
        this.parameterAnnotations = parameterAnnotations;
    }

    public String getCommand() {
        return command;
    }

    public String completeCommand(String path) {
        return completePath(path, command);
    }

    public Class<?>[] getParamTypes() {
        return paramTypes.clone();
    }

    public short getPriority() {
        return priority;
    }

    public String getHelpMessage() {
        return helpMessage;
    }

    public boolean isCaseSensitive() {
        return caseSensitive;
    }

    public CommandHandlerInternal getHandler() {
        return handler;
    }

    public boolean handle(Player player, Object[] params) {
        return handler.handle(player, params);
    }

    public String getCategorie() {
        return categorie;
    }

    public List<CustomCommandHandler> getBeforeCheckers() {
        return beforeCheckers;
    }

    public List<CustomCommandHandler> getCustomHandlers() {
        return customHandlers;
    }

    public CommandParameter[] getParameterAnnotations() {
        return parameterAnnotations;
    }
}

File: src/main/java/net/gtaun/shoebill/common/command/CommandEntryInternal.kt
package net.gtaun.shoebill.common.command

import net.gtaun.shoebill.entities.Player
import org.apache.commons.lang3.StringUtils

class CommandEntryInternal internal constructor(val command: String, val paramTypes: Array<Class<*>>, val priority: Short,
                                                val helpMessage: String?, val isCaseSensitive: Boolean,
                                                val handler: CommandEntryInternal.CommandHandlerInternal,
                                                val category: String?, val beforeCheckers: List<CustomCommandHandler>,
                                                val customHandlers: List<CustomCommandHandler>,
                                                val parameterAnnotations: Array<CommandParameter> = arrayOf()) {

    @FunctionalInterface
    interface CommandHandlerInternal {
        fun handle(player: Player, params: Array<Any>): Boolean
    }

    fun completeCommand(path: String): String = completePath(path, command)

    fun handle(player: Player, params: Array<Any>): Boolean = handler.handle(player, params)

    companion object {
        @JvmStatic
        fun completePath(path: String, child: String): String =
                if (StringUtils.isBlank(path)) child else path + " " + child

        fun CommandHandlerInternal(handler: (Player, Array<Any>) -> Boolean) = object : CommandHandlerInternal {
            override fun handle(player: Player, params: Array<Any>): Boolean = handler(player, params)
        }
    }
}
